azerothcore/azerothcore-wotlk
# feat(CORE/Instance): access_requirement db refactor and improved output

## CHANGES PROPOSED:
-  Changed the DB to allow multiple requirements for different types
-  Added new output to let players know what is missing to gain access including clickable links for items, quests and achievements
- Change old command for the changes
- Added configuration to chose the correct output to print
Thanks @BarbzYHOOL for sql provided!


## ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/3644
- Close https://github.com/azerothcore/azerothcore-wotlk/issues/2832 and https://github.com/azerothcore/azerothcore-wotlk/issues/2815



## TESTS PERFORMED:
- Imported Db
- Tried to enter the dungeon without meeting and meeting the requirements
![image](https://user-images.githubusercontent.com/4162295/99159047-f7f4ca00-26d0-11eb-879c-a91d0780a20a.png)



## HOW TO TEST THE CHANGES:

1. Import new DB and compile
2. Go to any dungeon that has requirements for quest/items or other
3. Try to enter without the required quests/items
4. Try to enter with the requirements
5. Try dungeon finder and see if any dungeon is locked incorrectly.
A good Dungeon to try is The Black morass in Cavern of Times (Tanaris) with a new character as it requieres a quest to complete before you can enter.

Use this SQL to add even more requirements to The Black Morass:
```
INSERT INTO `dungeon_access_requirements` (`dungeon_access_id`, `requirement_type`, `requirement_id`) VALUES ('17', '2', '50736');
INSERT INTO `dungeon_access_requirements` (`dungeon_access_id`, `requirement_type`, `requirement_id`) VALUES ('17', '0', '3579');
INSERT INTO `dungeon_access_requirements` (`dungeon_access_id`, `requirement_type`, `requirement_id`, `requirement_hint`) VALUES ('17', '0', '3580', 'Testo hinto for food fight');
```
This adds 2 achievements requierements and 1 item requirements that can be then adquired in game using:
.achive add 3579
.achive add 3580
.additem 50736

## KNOWN ISSUES AND TODO LIST:

- [ ] Update wiki

## Target branch(es):








 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR
